object/relations/relations.go

```diff
@@ -126,3 +129,230 @@ func ListRelations(ctx context.Context, rels Relations, cnrID cid.ID, rootObjID
 
 	return chain, nil
 }
+
+// HeadExecutor describes methods to get object head.
+type HeadExecutor interface {
+	ObjectHead(ctx context.Context, containerID cid.ID, objectID oid.ID, prm client.PrmObjectHead) (*client.ResObjectHead, error)
+}
+
+// SearchExecutor describes methods to search objects.
+type SearchExecutor interface {
+	ObjectSearchInit(ctx context.Context, containerID cid.ID, prm client.PrmObjectSearch) (*client.ObjectListReader, error)
+}
+
+// Executor describes all methods required to find all siblings for object.
+type Executor interface {
+	HeadExecutor
+	SearchExecutor
+}
+
+// All returns all related phy objects for provided root object ID in split-chain order.
+// If linking object is found its id will be the last one.
+//
+// Result doesn't include root object ID itself.
+func All(ctx context.Context, executor Executor, cnrID cid.ID, rootObjID oid.ID, tokens Tokens) ([]oid.ID, error) {
+	return WithLinking(ctx, executor, cnrID, rootObjID, tokens, true)
+}
+
+// WithLinking returns all related phy objects for provided root object ID in split-chain order.
+//
+// Result doesn't include root object ID itself.
+func WithLinking(ctx context.Context, executor Executor, containerID cid.ID, rootObjectID oid.ID, tokens Tokens, includeLinking bool) ([]oid.ID, error) {
+	splitInfo, err := getSplitInfo(ctx, executor, containerID, rootObjectID, tokens)
+	if err != nil {
+		if errors.Is(err, ErrNoSplitInfo) {
+			return []oid.ID{}, nil
+		}
+
+		return nil, err
+	}
+
+	// collect split chain by the descending ease of operations (ease is evaluated heuristically).
+	// If any approach fails, we don't try the next since we assume that it will fail too.
+	if _, ok := splitInfo.Link(); !ok {
+		// the list is expected to contain last part and (probably) split info
+		list, err := findSiblingByParentID(ctx, executor, containerID, rootObjectID, tokens)
+		if err != nil {
+			return nil, fmt.Errorf("children: %w", err)
+		}
+
+		for _, id := range list {
+			split, err := getSplitInfo(ctx, executor, containerID, id, tokens)
+			if err != nil {
+				if errors.Is(err, ErrNoSplitInfo) {
+					continue
+				}
+				return nil, fmt.Errorf("split info: %w", err)
+			}
+			if link, ok := split.Link(); ok {
+				splitInfo.SetLink(link)
```

Comment:
why not `break` here and below (if last is found)?

Reply:
Don't we need to find both of them simultaneously? I mean if any of it is found we should stop our search?

Reply:
oh, you have just copy-pasted it from the previous implementation, did not mention that

> I mean if any of it is found we should stop our search?

i reacted to that line too hastily. paraphrase my comment: if a linking object is found, you just `HEAD` it, and the work is done (so no need to look for a last one): `break` can be called. if the last is found we keep searching for a linking object

Reply:
Got it, updated

Reply:
hasn't pushed yet?

Reply:
Yes, was something with commits, updated one more time